### PR TITLE
Add new TB Example for Dynamic ROM Creation

### PR DESCRIPTION
**Pull request by SushobhitGupt**
adding data extraction for a new example under construction

### CODE REVIEW

File: _unittest/test_00_downloads.py [hunk 1]
```diff
@@ -54,3 +54,7 @@ def test_download_custom_report(self):
     def test_download_3dcomp(self):
         out = self.examples.download_3dcomponent()
         assert os.path.exists(out)
+
+    def test_download_twin_builder_data(self):
+        out = self.examples.download_twin_builder_data("Ex1_Mechanical_DynamicRom.zip", True)
+        assert os.path.exists(out)
```
Comment (MaxJPRey) [inline _unittest/test_00_downloads.py:60]:
```suggestion
        example_folder = self.examples.download_twin_builder_data("Ex1_Mechanical_DynamicRom.zip", True)
        assert os.path.exists(example_folder)
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+###############################################################################
+# Schematic Creation
+# ~~~~~~~~~~~~~~~~~~
+# Place components to create a schematic.
+ 
+# Define the grid distance for ease in calculations.
+
+G = 0.00254
+
+# Place an Dynamic ROM component.
```
Comment (MaxJPRey) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:100]:
```suggestion
# Place a Dynamic ROM component.
```

File: pyaedt/examples/downloads.py [hunk 3]
```diff
@@ -554,6 +554,44 @@ def download_multiparts(destination=None):
...
+
+    Examples files are downloaded to a persistent cache to avoid
+    re-downloading the same file twice.
+
+    Parameters
+    ----------
+    file_name : str
+        Path of the file in Twin Builder folder.
+    force_download : bool, optional
+        Force to delete file and download file again.
```
Comment (MaxJPRey) [inline pyaedt/examples/downloads.py:568]:
```suggestion
        Force to delete file and download file again. Default value is ``False``.
```

File: pyaedt/examples/downloads.py [hunk 4]
```diff
@@ -554,6 +554,44 @@ def download_multiparts(destination=None):
...
+    'C:/Users/user/AppData/local/temp/twin_builder'
+    """
+    if not destination:
+        destination = EXAMPLES_PATH
+    if force_download:
+        local_path = os.path.join(destination, os.path.join("twin_builder", file_name))
+        if os.path.exists(local_path):
+            os.unlink(local_path)
+    _download_file("twin_builder", file_name, destination)
+    return os.path.join(destination, "twin_builder")
```
Comment (MaxJPRey) [inline pyaedt/examples/downloads.py:592]: @SushobhitGupt Looks great. Thanks for adding the detailed example.

**Comment by MaxJPRey** [comment]
@SushobhitGupt If we want this example to be available for internal testing, we could just let the branch being and not merge it in main yet.
Then, to install this "beta" version,  a user could do:
`pip install https://github.com/pyansys/pyaedt.git@main_sgupt`

This is the best practice and usual way to do it when someone wants to introduce a beta feature.
Here it is very specific as it is an example.

**Comment by SushobhitGupt** [comment]
> @SushobhitGupt If we want this example to be available for internal testing, we could just let the branch being and not merge it in main yet. Then, to install this "beta" version, a user could do: `pip install https://github.com/pyansys/pyaedt.git@main_sgupt`
> 
> This is the best practice and usual way to do it when someone wants to introduce a beta feature. Here it is very specific as it is an example.

@MaxJPRey Thanks for looking at it, one of the challenges of keeping items on personal branches is that the visibility and access to functionality gets limited. Other than a few folks, not many others have enough git knowledge to effectively use it. 

We are hoping to use this and other examples I plan to add to market the ease of automation using pyaedt in our release material even during the Beta period. 

I was hoping to point my ace and customers to examples on the web site while they play with our daily / beta builds. On the product side in AEDT, we generally merge beta functionality and just put a reasonable disclaimer on it. Let me know what you think?

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
+"""
+Twin Builder: Dynamic ROM Creation and Simulation (2023 R1 Beta)
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:2]:
```suggestion
Twin Builder: dynamic ROM creation and simulation (2023 R1 beta)
```

Comment (PipKat) [reply examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:2]: Goggle style uses snake case in titles and headings.

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+---------------------------------------------------------
+This example shows how you can use PyAEDT to create a Dynamic ROM in Twin Builder
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:4]:
```suggestion
This example shows how you can use PyAEDT to create a dynamic ROM in Twin Builder
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+and run a Twin Builder time-domain simulation.
+
+Usage Notice:
+This example uses functionality only available in Twin Builder 2023 R1 or higher
+with the minimum build date of 8/7/2022
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:9]:
```suggestion
.. note::
    This example uses functionality only available in Twin Builder 2023 R1 and later.
    For 2023 R1, the build date must be 8/7/2022 or later. 
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+"""
+
+###############################################################################
+# Perform required imports
+# ~~~~~~~~~~~~~~~~~~~~~~~
+# Perform the required imports.
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:15]:
```suggestion
# Perform required imports.
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+# Twin Builder in non-graphical mode. You can also change the Boolean parameter
+# ``new_thread`` to ``False`` to launch Twin Builder in an existing AEDT session
+# if one is running.
+
+desktop_version = "2023.1"
+non_graphical = os.getenv("PYAEDT_NON_GRAPHICAL", "False").lower() in ("true", "1", "t")
+new_thread = True
+
+###############################################################################
+# Setup Input Data
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:42]:
```suggestion
# Set up input data
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+# ~~~~~~~~~~~~~~~~
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:43]:
```suggestion
# ~~~~~~~~~~~~~~~~~
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+
+data_folder = os.path.join(source_data_folder, "Ex03")
+
+# Unzip training data and config file
+examples.unzip(os.path.join(source_data_folder ,source_snapshot_data_zipfilename), data_folder)
+shutil.copyfile(os.path.join(source_data_folder ,source_build_conf_file), os.path.join(data_folder,source_build_conf_file))
+
+
+###############################################################################
+# Launch Twin Builder and build the Dynamic ROM Component
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:65]:
```suggestion
# Launch Twin Builder and build the dynamic ROM component
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+# Launch Twin Builder using an implicit declaration and add a new design with
+# a default setup.
+
+tb = TwinBuilder(projectname=generate_unique_project_name(),specified_version=desktop_version, non_graphical=non_graphical, new_desktop_session=new_thread)
+
+# Get the Dynamic ROM Builder object
+romMgr = tb._odesign.GetROMManager()
+dromBuilder = romMgr.GetDynamicROMBuilder()
+
+# Build the Dynamic ROM with specified configuration file
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:76]:
```suggestion
# Build the dynamic ROM with specified configuration file
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+else:
+	tb._odesign.AddMessage("Info","path does not exist: {}".format(drompath), "")
+
+#Create the ROM component definition in Twin Builder
+romMgr.CreateROMComponent(drompath.replace('\\', '/'),'dynarom') 
+
+
+###############################################################################
+# Schematic Creation
+# ~~~~~~~~~~~~~~~~~~
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:93]:
```suggestion
# Create schematic
# ~~~~~~~~~~~~~~
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+# Place components to create a schematic.
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:94]:
```suggestion
# Place components to create a schematic
```

Comment (PipKat) [reply examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:94]: Within this example's code blocks, the majority of comments do not have periods. I don't care if they do or don't, be I do care that they are consistent.

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+ 
+# Define the grid distance for ease in calculations.
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:96]:
```suggestion
# Define the grid distance for ease in calculations
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+G = 0.00254
+
+# Place an Dynamic ROM component.
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py]:
```suggestion
# Place a dynamic ROM component
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+
+rom1 = tb.modeler.schematic.create_component("ROM1","","dynarom", [36 * G, 28 * G])
+
+# Place two Excitation sources.
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:104]:
```suggestion
# Place two excitation sources.
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+
+source1 = tb.modeler.schematic.create_periodic_waveform_source(None, "PULSE", 190, 0.002, "300deg", 210, 0, [20 * G, 29 * G])
+source2 = tb.modeler.schematic.create_periodic_waveform_source(None, "PULSE", 190, 0.002, "300deg", 210, 0, [20 * G, 25 * G])
+
+# Connect components with wires.
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:109]:
```suggestion
# Connect components with wires
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+
+tb.analyze_setup("TR")
+
+
+###############################################################################
+# Get report data and plot using Matplotlib
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Get report data and plot it using Matplotlib. The following code gets and plots
+# the values for the voltage on the pulse voltage source and the values for the
+# output of the Dynamic ROM.
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:139]:
```suggestion
# output of the dynamic ROM.
```

File: pyaedt/examples/downloads.py [hunk 20]
```diff
@@ -565,7 +565,7 @@ def download_twin_builder_data(file_name, force_download=False, destination=None
     file_name : str
         Path of the file in Twin Builder folder.
```
Comment (PipKat) [inline pyaedt/examples/downloads.py:566]:
```suggestion
        Path of the file in the Twin Builder folder.
```

File: pyaedt/examples/downloads.py [hunk 21]
```diff
@@ -565,7 +565,7 @@ def download_twin_builder_data(file_name, force_download=False, destination=None
...
     force_download : bool, optional
-        Force to delete cache and download files again.
+        Force to delete file and download file again.
```
Comment (PipKat) [inline pyaedt/examples/downloads.py]:
```suggestion
        Whether to force the file to be deleted and downloaded again. The default is ``False``.
```

File: pyaedt/examples/downloads.py [hunk 22]
```diff
@@ -565,7 +565,7 @@ def download_twin_builder_data(file_name, force_download=False, destination=None
...
     destination : str, optional
         Path where files will be downloaded. Optional. Default is user temp folder.
```
Comment (PipKat) [inline pyaedt/examples/downloads.py]:
```suggestion
        Path to download files to. The default is your temporary folder.
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py (hunk 5 above)
Comment (SushobhitGupt) [reply examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:2]: @PipKat Thanks for letting me know

File: pyaedt/examples/downloads.py [hunk 23]
```diff
@@ -554,6 +554,44 @@ def download_multiparts(destination=None):
...
+    re-downloading the same file twice.
+
+    Parameters
+    ----------
+    file_name : str
+        Path of the file in the Twin Builder folder.
+    force_download : bool, optional
+        Force to delete file and download file again. Default value is ``False``.
+    destination : str, optional
+        Path where files will be downloaded. Default is ``user temp`` folder.
```
Comment (PipKat) [inline pyaedt/examples/downloads.py:570]: Is the name of the folder ``user temp``? If not, we don't want to use literal tags. Based on the example below, It should be "The default is the user's temporary folder." (I see what you don't want to use the pronoun "your" here.)

Comment (SushobhitGupt) [reply pyaedt/examples/downloads.py:570]: @PipKat This parameter is taken verbatim from other similar functions in this file. 
They mention the user temp folder

Comment (PipKat) [inline pyaedt/examples/downloads.py:570]:
```suggestion
        Path to download files to. The default is the user's temporary folder.
```

File: pyaedt/examples/downloads.py [hunk 24]
```diff
@@ -554,6 +554,44 @@ def download_multiparts(destination=None):
     return os.path.join(dest_folder, "library")
 
 
+def download_twin_builder_data(file_name, force_download=False, destination=None):
+    """Download specified Twin Builder example data file.
```
Comment (PipKat) [inline pyaedt/examples/downloads.py:558]:
```suggestion
    """Download Twin Builder example data files.
```

File: pyaedt/examples/downloads.py [hunk 25]
```diff
@@ -554,6 +554,44 @@ def download_multiparts(destination=None):
...
+
+    Examples files are downloaded to a persistent cache to avoid
+    re-downloading the same file twice.
```
Comment (PipKat) [inline pyaedt/examples/downloads.py:561]:
```suggestion
    downloading the same file twice.
```

File: pyaedt/examples/downloads.py (hunk 23 above)
Comment (PipKat) [inline pyaedt/examples/downloads.py:570]:
```suggestion
    file_name : str
        Path for the file in the Twin Builder folder.
    force_download : bool, optional
        Whether to force deletion of a downloaded file and download it again. The
        default is ``False``.
    destination : str, optional
        Path to download the file to. The default is the user's temporary folder.
```

File: pyaedt/examples/downloads.py [hunk 26]
```diff
@@ -554,6 +554,44 @@ def download_multiparts(destination=None):
...
+        Path of the file in the Twin Builder folder.
+    force_download : bool, optional
+        Force to delete file and download file again. Default value is ``False``.
+    destination : str, optional
+        Path where files will be downloaded. Default is ``user temp`` folder.
+
+    Returns
+    -------
+    str
+        Path to the example folder containing all example files.
```
Comment (PipKat) [inline pyaedt/examples/downloads.py:575]:
```suggestion
        Path to the folder containing all Twin Builder example data files.
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+data_folder = os.path.join(source_data_folder, "Ex03")
+
+# Unzip training data and config file
+examples.unzip(os.path.join(source_data_folder ,source_snapshot_data_zipfilename), data_folder)
+shutil.copyfile(os.path.join(source_data_folder ,source_build_conf_file), os.path.join(data_folder,source_build_conf_file))
+
+
+###############################################################################
+# Launch Twin Builder and build the dynamic ROM component
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:66]:
```suggestion
# Launch Twin Builder and build ROM component
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 28]
```diff
@@ -0,0 +1,165 @@
...
+# Launch Twin Builder using an implicit declaration and add a new design with
+# a default setup.
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:68]:
```suggestion
# a default setup for building the dynamic ROM component.
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 29]
```diff
@@ -0,0 +1,165 @@
...
+
+tb = TwinBuilder(projectname=generate_unique_project_name(),specified_version=desktop_version, non_graphical=non_graphical, new_desktop_session=new_thread)
+
+# Get the Dynamic ROM Builder object
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:72]:
```suggestion
# Get the dynamic ROM builder object
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 30]
```diff
@@ -0,0 +1,165 @@
...
+	tb._odesign.AddMessage("Info","path does not exist: {}".format(drompath), "")
+
+#Create the ROM component definition in Twin Builder
+romMgr.CreateROMComponent(drompath.replace('\\', '/'),'dynarom') 
+
+
+###############################################################################
+# Create schematic
+# ~~~~~~~~~~~~~~~~
+# Place components to create a schematic
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:94]:
```suggestion
# Place components to create a schematic.
```

File: examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py [hunk 31]
```diff
@@ -0,0 +1,165 @@
...
+ 
+# Define the grid distance for ease in calculations
+
+G = 0.00254
+
+# Place a Dynamic ROM component.
```
Comment (PipKat) [inline examples/07-TwinBuilder/03-Dynamic_ROM_Creation_And_Visualization.py:100]:
```suggestion
# Place a dynamic ROM component.
```

**Review by MaxJPRey — APPROVED** [review]

**Comment by MaxJPRey** [comment]
LGTM.

**Pull request merged**